src/librustc_typeck/astconv.rs

```diff
@@ -1531,50 +1528,82 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+
+                                          traits::transitive_bounds(tcx, bounds)
+                                      },
                                       &param_name.as_str(),
                                       assoc_name,
                                       span)
     }
 
-    // Checks that `bounds` contains exactly one element and reports appropriate
+    // Checks that `all_candidates` contains exactly one matching element and reports appropriate
     // errors otherwise.
     fn one_bound_for_assoc_type<I>(&self,
-                                   mut bounds: I,
+                                   all_candidates: impl Fn() -> I,
                                    ty_param_name: &str,
                                    assoc_name: ast::Ident,
                                    span: Span)
         -> Result<ty::PolyTraitRef<'tcx>, ErrorReported>
         where I: Iterator<Item = ty::PolyTraitRef<'tcx>>
     {
-        let bound = match bounds.next() {
+        let mut matching_candidates = all_candidates().filter(|r| {
+            self.trait_defines_associated_type_named(r.def_id(), assoc_name)
+        });
+
+        let bound = match matching_candidates.next() {
             Some(bound) => bound,
             None => {
-                struct_span_err!(self.tcx().sess, span, E0220,
+                let mut err = struct_span_err!(self.tcx().sess, span, E0220,
                                  "associated type `{}` not found for `{}`",
                                  assoc_name,
-                                 ty_param_name)
-                    .span_label(span, format!("associated type `{}` not found", assoc_name))
-                    .emit();
+                                 ty_param_name);
+
+                let all_candidate_names: Vec<_> = all_candidates()
+                    .map(|r| self.tcx().associated_items(r.def_id()))
+                    .flatten()
+                    .filter_map(|item|
+                        if item.kind == ty::AssocKind::Type {
+                            Some(item.ident.name)
+                        } else {
+                            None
+                        }
+                    )
+                    .collect();
+
+                if let Some(suggested_name) = find_best_match_for_name(
+                    all_candidate_names.iter(),
+                    &assoc_name.as_str(),
+                    None,
+                ) {
+                    err.span_suggestion(
+                        span,
+                        "there is an associated type with a similar name",
+                        suggested_name.to_string(),
+                        Applicability::MaybeIncorrect,
+                    );
+                } else {
+                    err.span_label(
+                        span,
+                        format!("associated type `{}` not found", assoc_name)
+                    );
+                }
+
+                err.emit();
```

Comment:
Could we move all this to a separate method called `complain_about_assoc_type_not_found`?